### PR TITLE
RavenDB-11135 Cluster transactions should be supported in backup, sna…

### PR DESCRIPTION
**Pull request by karmeli87**
…pshot and smuggler

### CODE REVIEW

File: src/Raven.Server/Documents/DatabasesLandlord.cs [hunk 1]
```diff
@@ -206,6 +206,13 @@ private void NotifyPendingClusterTransaction(string databaseName, Task<DocumentD
                                         Array = mergedCommands.Reply,
                                         IndexTask = indexTask,
                                     };
+
+                                    if (mergedCommands.Options?.TaskId == null)
```
Comment (ayende) [inline src/Raven.Server/Documents/DatabasesLandlord.cs]: How can this be null? Because if so, in line 216 we have NRE

File: src/Raven.Server/Documents/DocumentPutAction.cs [hunk 2]
```diff
@@ -115,6 +115,12 @@ public DocumentPutAction(DocumentsStorage documentsStorage, DocumentDatabase doc
 
                 var result = BuildChangeVectorAndResolveConflicts(context, id, lowerId, newEtag, document, changeVector, expectedChangeVector, flags, oldValue);
 
+                if (flags.Contain(DocumentFlags.FromClusterTransaction) && nonPersistentFlags.Contain(NonPersistentDocumentFlags.FromSmuggler))
+                {
+                    var index = ChangeVectorUtils.GetEtagById(context.LastDatabaseChangeVector, _documentDatabase.DatabaseGroupId);
+                    changeVector = $"RAFT:{index}-{_documentDatabase.DatabaseGroupId}";
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentPutAction.cs]: This is bad, because it doesn't take into account the cluster level index.
A cluster tx on this may very well have a raft index that is smaller than the etag, resulting in us expecting that this is already committed and ignoring the change.

File: src/Raven.Client/Documents/Session/ClusterTransactionSession.cs [hunk 3]
```diff
@@ -85,6 +85,8 @@ public void Clear()
             _storeCompareExchange = null;
         }
 
+        public bool HasCommands => _deleteCompareExchange != null || _storeCompareExchange != null;
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/ClusterTransactionSession.cs]: internal?

File: src/Raven.Client/Documents/Smuggler/DatabaseSmuggler.cs [hunk 4]
```diff
@@ -125,10 +125,13 @@ public static DatabaseItemType ConfigureOptionsForIncrementalImport(DatabaseSmug
             // as the previous files can hold indexes and identities which were deleted and shouldn't be imported
             var oldOperateOnTypes = options.OperateOnTypes;
             options.OperateOnTypes = options.OperateOnTypes &
-                                     ~(DatabaseItemType.Indexes | DatabaseItemType.CompareExchange | DatabaseItemType.Identities);
+                                     ~(DatabaseItemType.Indexes | ClusterStateMachineImportOptions);
             return oldOperateOnTypes;
         }
 
+        public static DatabaseItemType ClusterStateMachineImportOptions =>
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Smuggler/DatabaseSmuggler.cs]: private or internal

File: src/Raven.Server/ServerWide/Commands/CompareExchangeCommands.cs [hunk 5]
```diff
@@ -22,7 +22,7 @@ public abstract class CompareExchangeCommandBase : CommandBase
 
         public static string GetActualKey(string database, string key)
         {
-            return ("db/" + database + "/" + key).ToLowerInvariant();
+            return (database + "/" + key).ToLowerInvariant();
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/Commands/CompareExchangeCommands.cs:25]: isn't this breaking backward-compatibility?

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 6]
```diff
@@ -254,11 +259,62 @@ private void FinishBatchOfDocuments()
             }
         }
 
+        private class DatabasePendingClusterTransacinsActions : IValueActions<ClusterTransactionCommand.SingleClusterDatabaseCommand>
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: typo :)

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 7]
```diff
@@ -254,11 +259,62 @@ private void FinishBatchOfDocuments()
...
+                if (_commands.Count == 0)
+                    return;
+
+                SendTransactionCommands();
+            }
+
+            private void SendTransactionCommands()
+            {
+                var batch = new ClusterTransactionBatchCommand(_commands);
+                AsyncHelpers.RunSync(async () => await _database.ServerStore.SendToLeaderAsync(batch));
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: no need to add async/await here?

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 8]
```diff
@@ -268,25 +324,27 @@ public DatabaseCompareExchangeActions(DocumentDatabase database, JsonOperationCo
...
 
-                SendCommands(_context);
+                SendCommands();
             }
 
-            private void SendCommands(JsonOperationContext context)
+            private void SendCommands()
             {
-                AsyncHelpers.RunSync(async () => await _database.ServerStore.SendToLeaderAsync(new AddOrUpdateCompareExchangeBatchCommand(_compareExchangeCommands, context)));
+                AsyncHelpers.RunSync(async () => await _database.ServerStore.SendToLeaderAsync(new AddOrUpdateCompareExchangeBatchCommand(_compareExchangeCommands, _context)));
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: can remove async/await from here?

File: src/Sparrow/Json/BlittableJsonTextWriter.cs [hunk 9]
```diff
@@ -137,6 +138,11 @@ public override string ToString()
             return Encodings.Utf8.GetString(_pinnedBuffer.Buffer.Array, _pinnedBuffer.Buffer.Offset, _pos);
         }
 
+        public void WriteIDynamicJsonWithContext(IDynamicJsonWithContext obj)
```
Comment (ppekrol) [inline src/Sparrow/Json/BlittableJsonTextWriter.cs]: WriteObject?

File: src/Raven.Server/Smuggler/Documents/StreamSource.cs [hunk 10]
```diff
@@ -225,6 +228,43 @@ public IDisposable GetCounterValues(out IEnumerable<CounterDetail> counters)
...
+        }
+
+        private IEnumerable<ClusterTransactionCommand.SingleClusterDatabaseCommand> InternalGetPendingClusterTransactions(DisposableScope scope)
+        {
+            foreach (BlittableJsonReaderObject reader in ReadArray(disposableScope: scope))
+            {
+                scope.EnsureDispose(reader);
+                if (reader.TryGet(nameof(ClusterTransactionCommand.SingleClusterDatabaseCommand.Commands), out BlittableJsonReaderArray commandsArray) == false)
+                {
+                    _result.CompareExchange.ErroredCount++;
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/StreamSource.cs]: CompareExchange?

File: src/Raven.Client/Documents/Smuggler/DatabaseItemType.cs [hunk 11]
```diff
@@ -18,6 +18,7 @@ public enum DatabaseItemType
         LegacyDocumentDeletions = 1 << 9,
         LegacyAttachmentDeletions = 1 << 10,
         DatabaseRecord = 1 << 11,
-        Counters = 1 << 12
+        Counters = 1 << 12,
+        PendingClusterTransactions = 1 << 13
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Smuggler/DatabaseItemType.cs:22]: just 'ClusterTransactions' ?

**Comment by ppekrol** [comment]
@karmeli87 please check the http://jenkins.hibernatingrhinos.com/job/PR_Conventions/1535/console

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 12]
```diff
@@ -1140,7 +1150,8 @@ public IEnumerable<(string Key, long Index)> GetCompareExchangeIndexes(Transacti
         private static unsafe string ReadCompareExchangeKey(TableValueReader reader, string dbPrefix)
         {
             var ptr = reader.Read((int)UniqueItems.Key, out var size);
-            return Encodings.Utf8.GetString(ptr, size).Substring(dbPrefix.Length + 4);
+            // we need to read only the key from the format: 'databaseName/key'
+            return Encodings.Utf8.GetString(ptr, size).Substring(dbPrefix.Length + 1);
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs:1154]: why this change now?

File: src/Raven.Client/Documents/Session/ClusterTransactionSession.cs (hunk 3 above)
Comment (karmeli87) [reply src/Raven.Client/Documents/Session/ClusterTransactionSession.cs]: done

File: src/Raven.Server/ServerWide/Commands/CompareExchangeCommands.cs (hunk 5 above)
Comment (ayende) [inline src/Raven.Server/ServerWide/Commands/CompareExchangeCommands.cs:25]: This is to match 4.0 behavior, right?

File: src/Raven.Client/Documents/Smuggler/DatabaseSmuggler.cs (hunk 4 above)
Comment (karmeli87) [reply src/Raven.Client/Documents/Smuggler/DatabaseSmuggler.cs]: done

File: src/Raven.Server/Smuggler/Documents/DatabaseSource.cs [hunk 13]
```diff
@@ -248,6 +252,18 @@ public IDisposable GetCompareExchangeValues(out IEnumerable<(string key, long in
             }
         }
 
+        public IDisposable GetPendingClusterTransactions(out IEnumerable<ClusterTransactionCommand.SingleClusterDatabaseCommand> transaction)
+        {
+            using (var scope = new DisposableScope())
+            {
+                scope.EnsureDispose(_database.ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context));
+                scope.EnsureDispose(context.OpenReadTransaction());
+                // if we export the transaction without having the document, we will not check if the transaction was already applied.
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/DatabaseSource.cs:261]: What is the meaning here?

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/ServerWide/Commands/CompareExchangeCommands.cs (hunk 5 above)
Comment (karmeli87) [reply src/Raven.Server/ServerWide/Commands/CompareExchangeCommands.cs:25]: The previous implementation, broke it. This should fix it :)

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs (hunk 6 above)
Comment (karmeli87) [reply src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: done

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs (hunk 7 above)
Comment (karmeli87) [reply src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: done

File: src/Raven.Server/Documents/DatabasesLandlord.cs (hunk 1 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/DatabasesLandlord.cs]: done

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs (hunk 12 above)
Comment (karmeli87) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs:1154]: Because we changed the key format. (removed the 'db/' prefix)

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs (hunk 8 above)
Comment (karmeli87) [reply src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: done

File: src/Raven.Server/Smuggler/Documents/DatabaseSource.cs (hunk 13 above)
Comment (karmeli87) [reply src/Raven.Server/Smuggler/Documents/DatabaseSource.cs:261]: When we export transaction _with_ documents we will check whether the transaction was already executed and if so, we will skip exporting it.
If we export _without_ the documents we will always take the stored transaction. (Not sure if there is a real case when we want to do something like this, but it is necessary for the tests)

File: src/Raven.Server/Smuggler/Documents/StreamSource.cs (hunk 10 above)
Comment (karmeli87) [reply src/Raven.Server/Smuggler/Documents/StreamSource.cs]: done

File: src/Sparrow/Json/BlittableJsonTextWriter.cs (hunk 9 above)
Comment (karmeli87) [reply src/Sparrow/Json/BlittableJsonTextWriter.cs]: done

File: src/Raven.Server/Documents/DocumentPutAction.cs (hunk 2 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/DocumentPutAction.cs]: This is indeed bad, because two document will have the same change vector.

the RAFT here is confusing, it doesn't reflect the raft index of the cluster, rather it is an incremental etag for each PUT/DELETE command that was passed through the cluster transaction.
With the unique Id of the database group.

File: src/Raven.Client/Documents/Smuggler/DatabaseItemType.cs (hunk 11 above)
Comment (ayende) [reply src/Raven.Client/Documents/Smuggler/DatabaseItemType.cs:22]: That is actually pending ones, that haven't been applied yet

File: src/Raven.Server/Documents/DocumentPutAction.cs [hunk 14]
```diff
@@ -115,6 +115,17 @@ public DocumentPutAction(DocumentsStorage documentsStorage, DocumentDatabase doc
 
                 var result = BuildChangeVectorAndResolveConflicts(context, id, lowerId, newEtag, document, changeVector, expectedChangeVector, flags, oldValue);
 
+                if (flags.Contain(DocumentFlags.FromClusterTransaction) && nonPersistentFlags.Contain(NonPersistentDocumentFlags.FromSmuggler))
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentPutAction.cs:118]: Strip `FromClusterTransaction` during import

**Review by ayende — CHANGES_REQUESTED** [review]

**Comment by karmeli87** [comment]
Instead of this complex commit, we simply should wait on the export for all pending transactions to complete.
